### PR TITLE
Backport ao insert bug

### PR DESCRIPTION
**Pull request by berlin-ab**
A backport of PR #5792:

Fix a bug that caused new data to be inserted into segfiles in "awaiting drop" state

### CODE REVIEW

**Comment by ashwinstar** [comment]
Seems a lot of churn on code which could be easily cut-down if the commits were squashed properly. Definitely not a PR which can be reviewed commit by commit.

File: src/backend/access/appendonly/appendonlywriter.c [hunk 1]
```diff
@@ -58,9 +59,24 @@ static AORelHashEntry AppendOnlyRelHashNew(Oid relid, bool *exists);
 static AORelHashEntry AORelGetHashEntry(Oid relid);
 static AORelHashEntry AORelLookupHashEntry(Oid relid);
 static bool AORelCreateHashEntry(Oid relid);
-static bool *GetFileSegStateInfoFromSegments(Relation parentrel);
+static bool *get_awaiting_drop_status_from_segments(Relation parentrel);
 static int64 *GetTotalTupleCountFromSegments(Relation parentrel, int segno);
 
+static void
+acquire_lightweight_lock() {
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/backend/access/appendonly/appendonlywriter.c]: All references to the native call use the same arguments. If that native call changes in the future, all of those calls would need to change. Now, they only change in one place. Also, I no longer need to parse the arguments to the native methods to ensure that they're the same as before. That decision is made just once within `appendonlywriter.c`.

File: src/backend/access/appendonly/appendonlywriter.c [hunk 2]
```diff
@@ -2020,11 +2051,72 @@ AtEOXact_AppendOnly(void)
...
+}
+
+/*
+ * Remove an entry from the Append-only hash
+ *
+ * For external use. Use `AORelRemoveHashEntry`
+ * for internal removal of an AO hash entry.
+ */
+void
+GpRemoveEntryFromAppendOnlyHash(Oid relid) {
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/backend/access/appendonly/appendonlywriter.c]: I agree, a status would be better. I had transformed the outcome function pointers into local functions in response to @asimrp's comment on the original PR.

[the remark this answers is not in the document]
Comment (berlin-ab) [reply src/backend/access/appendonly/appendonlywriter.c]: We could always introduce that change as an improvement.

**Comment by berlin-ab** [comment]
I agree that this could have been squashed better before introducing into master. At the time, I thought that the commits each still held their own reasons for existing. During the backport I've noticed where there could be improvement.

**Pull request closed without merging**